common/src/primitives/height.rs

```diff
@@ -1,18 +1,83 @@
 use parity_scale_codec_derive::{Decode, Encode};
+use std::fmt;
+use std::ops::{Add, Sub};
+
+type HeightIntType = u64;
+type DistanceIntType = i64;
 
 #[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Encode, Decode)]
-pub struct BlockHeight(u64);
+pub struct BlockHeight(HeightIntType);
+
+// Display should be defined for thiserr crate
+impl fmt::Display for BlockHeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<BlockHeight> for HeightIntType {
+    fn from(block_height: BlockHeight) -> HeightIntType {
+        block_height.inner()
+    }
+}
+
+impl From<HeightIntType> for BlockHeight {
+    fn from(w: HeightIntType) -> BlockHeight {
+        BlockHeight(w)
+    }
+}
+
+impl Add<BlockDistance> for BlockHeight {
+    type Output = Option<Self>;
+
+    fn add(self, other: BlockDistance) -> Option<Self> {
+        let height: i64 = self.0 as i64;
+        if height < 0 {
+            return None;
+        }
```

Comment:
Could the cast-and-check pattern be replaced with something like the following?
```rust
let height = i64::try_from(self.0)?;
```
You may need to import the `TryFrom` trait, not sure if it's in scope by default.

Reply:
I'm sorry, I don't understand the difference compared to `let height: i64 = self.0.try_into().ok()?;`